types/q/index.d.ts

```diff
@@ -60,7 +60,7 @@ declare namespace Q {
 		/**
 		 * The then method from the Promises/A+ specification, with an additional progress handler.
 		 */
-		then<U>(onFulfill?: ((value: T) => IWhenable<U>) | null, onReject?: ((error: any) => IWhenable<U>) | null, onProgress?: ((progress: any) => any) | null): Promise<U>;
+		then<U, V = never>(onFulfill?: ((value: T) => IWhenable<U>) | null, onReject?: ((error: any) => IWhenable<V>) | null, onProgress?: ((progress: any) => any) | null): Promise<U>;
```

Comment:
Won't this break with the current version of typescript?

Reply:
Yes, it would, but that's kind of what the 'TypeScript Version' comment is for on line 9.  It would be nice if V defaulted to U, I'm guessing that would be the least surprising option for most people using Q.

Reply:
The `// TypeScript Version: 2.3` line tells the CI which minimum typescript version should be used for testing. If such a line is not present 2.0 is used.
To my understanding it is verified already that old typescript versions are not effected.

Therefore I think this doesn't break old typescript versions.

Please note that `PromiseLike.then` has not changed between [2.3](https://github.com/Microsoft/TypeScript/blob/v2.3.0/lib/lib.es5.d.ts#L1330) and [@next](https://github.com/Microsoft/TypeScript/blob/v3.1-rc/lib/lib.es5.d.ts#L1323). I think the change is in typescript itself to closer verify the signatures but I can't provide any link to a ts PR.

Reply:
@TeamworkGuy2 If I default V to U I get again a build error with @next:
```
ERROR: 7:11   expect  TypeScript@next compile error:
Argument of type '(x: number) => void' is not assignable to parameter of type '(value: number) => {} | PromiseLike<{}>'.
  Type 'void' is not assignable to type '{} | PromiseLike<{}>'.
ERROR: 55:9   expect  TypeScript@next compile error:
Argument of type '(results: [number, number]) => void' is not assignable to parameter of type '(value: [number, number]) => {} | PromiseLike<{}>'.
  Type 'void' is not assignable to type '{} | PromiseLike<{}>'.
ERROR: 68:8   expect  TypeScript@next compile error:
Argument of type '(value4: void) => void' is not assignable to parameter of type '(value: void) => {} | PromiseLike<{}>'.
  Type 'void' is not assignable to type '{} | PromiseLike<{}>'.
ERROR: 90:8   expect  TypeScript@next compile error:
Argument of type '(x: number) => void' is not assignable to parameter of type '(value: number) => {} | PromiseLike<{}>'.
  Type 'void' is not assignable to type '{} | PromiseLike<{}>'.
ERROR: 103:8  expect  TypeScript@next compile error:
Argument of type '{ (text: string): Promise<number>; (text: string): JQueryPromise<number>; }' is not assignable to parameter of type '(value: {}) => IWhenable<number>'.
  Types of parameters 'text' and 'value' are incompatible.
    Type '{}' is not assignable to type 'string'.
ERROR: 104:9  expect  TypeScript@next compile error:
Parameter 'num' implicitly has an 'any' type.
```